### PR TITLE
Fixing a few things for read/write syscalls and an uneeded abort decl…

### PR DESCRIPTION
**Pull request by elbing**
…aration

Signed-off-by: Álvaro Jurado <elbingmiss@gmail.com>

### CODE REVIEW

File: sys/src/9/port/sysfile.c [hunk 1]
```diff
@@ -717,11 +717,10 @@ mountfix(Chan *c, uint8_t *op, int32_t n, int32_t maxn)
 }
 
 static int32_t
-read(int fd, void *p, int32_t n, int64_t off)
+read(int ispread, int fd, void *p, int32_t n, int64_t off)
```
Comment (dancrossnyc) [inline sys/src/9/port/sysfile.c]: Why is this first for read, but last for write?

Comment (dancrossnyc) [reply sys/src/9/port/sysfile.c]: Also, I'd keep the 'sequential' name: it implies `ispread` would be true.

Comment (elbing) [reply sys/src/9/port/sysfile.c]: I just follow the original definition.

File: sys/src/9/port/sysfile.c [hunk 2]
```diff
@@ -742,21 +741,24 @@ read(int fd, void *p, int32_t n, int64_t off)
 	 * The number of bytes read on this fd (c->offset) may be different
 	 * due to rewritings in mountfix.
 	 */
+	if(ispread){	/* use and maintain channel's offset */
```
Comment (dancrossnyc) [inline sys/src/9/port/sysfile.c]: I think you can simplify this to,

    if(sequential || off == -1LL) {
        off = c->offset;
        sequential = 1;
    }

File: sys/src/9/port/sysfile.c [hunk 3]
```diff
@@ -742,21 +741,21 @@ read(int fd, void *p, int32_t n, int64_t off)
...
-		if(off == -1LL){	/* use and maintain channel's offset */
-			off = c->offset;
-		sequential = 1;
-	} else {
-		sequential = 0;
+	if(off == -1LL){	/* use and maintain channel's offset */
+		off = c->offset;
 	}
+	else
+		dopread = 0;
```
Comment (dancrossnyc) [inline sys/src/9/port/sysfile.c]: Isn't this reversed? If off == -1LL, then it's NOT pread (we're relying on the offset tracked in the Chan). If it's pread, then we use the offset passed as an argument.

File: sys/src/9/port/sysfile.c [hunk 4]
```diff
@@ -827,13 +825,11 @@ write(int fd, void *p, int32_t n, int64_t off)
 	n = 0;
 	c = fdtochan(fd, OWRITE, 1, 1);
 
-	if(off == -1LL)
-		sequential = 1;
-	else
-		sequential = 0;
+	if(off != -1LL)
+		dopwrite = 0;
```
Comment (dancrossnyc) [inline sys/src/9/port/sysfile.c]: Same here.

File: sys/src/9/port/sysfile.c [hunk 5]
```diff
@@ -847,7 +843,7 @@ write(int fd, void *p, int32_t n, int64_t off)
 
 	n = r;
 
-	if(sequential){	/* use and maintain channel's offset */
+	if(dopwrite){	/* use and maintain channel's offset */
```
Comment (dancrossnyc) [inline sys/src/9/port/sysfile.c]: Why not avoid advancing c->offset until after r->dev->write returns?

Comment (dancrossnyc) [reply sys/src/9/port/sysfile.c]: Answering my own question...If I had to guess:

The authors increment offset so that multiple system calls writing to the chan at the same time don't overlap. However, there's a time-of-check, time-of-use bug here (I think) with setting `off` when off==-1LL.

Probably the best thing is to add, `int pread = (off != -1LL);` at the beginning and test on that to set off from c->offset and increment c->offset before c->dev->write.

File: sys/src/9/port/sysfile.c [hunk 6]
```diff
@@ -856,7 +852,7 @@ write(int fd, void *p, int32_t n, int64_t off)
 
 	r = c->dev->write(c, p, n, off);
```
Comment (dancrossnyc) [inline sys/src/9/port/sysfile.c]: Hmm. `r` is signed. What if it's -1 here?

File: util/src/harvey/cmd/elf2c/elf2c.go [hunk 7]
```diff
@@ -139,7 +139,7 @@ func main() {
 		gencode(w, file, "code", mem, codestart, codeend)
 		gencode(w, file, "data", mem, datastart, dataend)
 	}
-	if err := ioutil.WriteFile(a[0], w.Bytes(), 0444); err != nil {
+	if err := ioutil.WriteFile(a[0], w.Bytes(), 0666); err != nil {
```
Comment (dancrossnyc) [inline util/src/harvey/cmd/elf2c/elf2c.go]: Is this related?

Comment (elbing) [reply util/src/harvey/cmd/elf2c/elf2c.go]: aaaah no, it came from the annoying rm init.h I have to make every time.

File: sys/src/9/port/sysfile.c (hunk 6 above)
Comment (elbing) [reply sys/src/9/port/sysfile.c]: suicide?

File: sys/src/9/port/sysfile.c [hunk 8]
```diff
@@ -742,21 +741,21 @@ read(int fd, void *p, int32_t n, int64_t off)
...
-		if(off == -1LL){	/* use and maintain channel's offset */
-			off = c->offset;
-		sequential = 1;
-	} else {
-		sequential = 0;
+	if(off == -1LL){	/* use and maintain channel's offset */
+		off = c->offset;
 	}
+	else
+		dopread = 1;
```
Comment (dancrossnyc) [inline sys/src/9/port/sysfile.c]: dopread is already 1?

File: sys/src/9/port/sysfile.c [hunk 9]
```diff
@@ -720,8 +720,7 @@ static int32_t
 read(int fd, void *p, int32_t n, int64_t off)
 {
 	Proc *up = externup();
-	int32_t nn, nnn;
-	int sequential;
+	int32_t nn, nnn, dopread = 1;
```
Comment (dancrossnyc) [inline sys/src/9/port/sysfile.c]: Add a line:

    int dopread = (off != -1LL);

File: sys/src/9/port/sysfile.c [hunk 10]
```diff
@@ -742,21 +741,21 @@ read(int fd, void *p, int32_t n, int64_t off)
 	 * The number of bytes read on this fd (c->offset) may be different
 	 * due to rewritings in mountfix.
 	 */
-		if(off == -1LL){	/* use and maintain channel's offset */
-			off = c->offset;
-		sequential = 1;
-	} else {
-		sequential = 0;
+	if(off == -1LL){	/* use and maintain channel's offset */
```
Comment (dancrossnyc) [inline sys/src/9/port/sysfile.c]: if (!dopread) off = c->offset;

File: sys/src/9/port/sysfile.c [hunk 11]
```diff
@@ -742,21 +741,21 @@ read(int fd, void *p, int32_t n, int64_t off)
...
 		/*
 		 * Directory read:
 		 * rewind to the beginning of the file if necessary;
 		 * try to fill the buffer via mountrockread;
-		 * set sequential to always maintain the Chan offset.
+		 * clear dopread to always maintain the Chan offset.
 		 */
-		if(off == 0LL){
-			if(sequential){
+		if(off == -1LL){
```
Comment (dancrossnyc) [inline sys/src/9/port/sysfile.c]: if(off == 0LL){
        c->offset = 0;
        c->devoffset = 0;
    }
    nnn = mountfix(c, p, nn, n);
    dopread = 0;

File: sys/src/9/port/sysfile.c [hunk 12]
```diff
@@ -775,12 +774,12 @@ read(int fd, void *p, int32_t n, int64_t off)
...
 		nnn = mountfix(c, p, nn, n);
 
-		sequential = 1;
+		dopread = 1;
 	}
 	else
 		nnn = nn = c->dev->read(c, p, n, off);
 
-	if(sequential){
+	if(dopread){
```
Comment (dancrossnyc) [inline sys/src/9/port/sysfile.c]: This is reversed; but I think you can just always update the offsets in every case.

    lock(&c->r.l);
    if(nn > 0)
        c->devoffset += nn;
    if (nnn > 0)
        c->offset += nnn;
    unlock(&c->r.l);

File: sys/src/9/port/sysfile.c [hunk 13]
```diff
@@ -827,13 +825,11 @@ write(int fd, void *p, int32_t n, int64_t off)
...
-	if(off == -1LL)
-		sequential = 1;
-	else
-		sequential = 0;
+	if(off != -1LL)
+		dopwrite = 1;
 
 	if(waserror()) {
-		if(sequential){
+		if(dopwrite){
```
Comment (dancrossnyc) [inline sys/src/9/port/sysfile.c]: if (!doprwrite)

`doprwrite==1` means that you are using `pwrite` semantics. The old `sequential` meant that you were doing `write` semantics.

File: sys/src/9/port/sysfile.c [hunk 14]
```diff
@@ -742,27 +742,24 @@ read(int fd, void *p, int32_t n, int64_t off)
...
 		 */
 		if(off == 0LL){
-			if(sequential){
-				c->offset = 0;
-				c->devoffset = 0;
-			}
-			mountrewind(c);
-			unionrewind(c);
-		}
+ 			if(notpread){
```
Comment (dancrossnyc) [inline sys/src/9/port/sysfile.c]: Looks like some random leading whitespace crept into this change?

Comment (elbing) [reply sys/src/9/port/sysfile.c]: Cleaned!
